src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -782,8 +783,24 @@ CoordinatedRemoteTransactionsPrepare(void)
 			continue;
 		}
 
-		StartRemoteTransactionPrepare(connection);
-		connectionList = lappend(connectionList, connection);
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, we start preparing the transaction, otherwise
+		 * we skip prepare since the connection didn't perform any write (read-only)
+		 */
+		dlist_iter placementIter;
+		dlist_foreach(placementIter, &connection->referencedPlacements)
```

Comment:
I would suggest adding a field to Connection struct, `hasProcessedWrite` and setting it to true in these places:

https://github.com/citusdata/citus/blob/94bc2703bce352fc5837ca1bda5b8fe65569097d/src/backend/distributed/connection/placement_connection.c#L459

https://github.com/citusdata/citus/blob/94bc2703bce352fc5837ca1bda5b8fe65569097d/src/backend/distributed/connection/placement_connection.c#L464

So basically `connection->hasProcessedWrite = true;` could be there, not `placementConnection->hasProcessedWrite`.

Then here the instead of `dlist_foreach` the check could become:

```c
if (connection->hasProcessedWrite) {
				StartRemoteTransactionPrepare(connection);
				connectionList = lappend(connectionList, connection);
}
```

Would this work?

Reply:
I'd avoid adding a new flag. Can we use `ConnectionModifiedPlacement()` function?